rp2040-hal/src/timer.rs

```diff
@@ -106,6 +112,75 @@ impl Timer {
 // safety: all write operations are synchronised and all reads are atomic
 unsafe impl Sync for Timer {}
 
+/// [`Timer`] based Delay
+pub struct Delay(());
+impl Delay {
+    fn delay_from_u32(&self, us: u32) {
+        // safety: only reads are made
+        let timer = unsafe { &*pac::TIMER::PTR };
+        let start = timer.timelr.read().bits();
+        while timer.timelr.read().bits().wrapping_sub(start) <= us {
```

Comment:
There is a slight risk that this misses the deadline and waits for too long:
If `us` is `u32::MAX`, the condition `now.wrapping_sub(start) <= us` is true for only 1µs. It is not unthinkable that the loop gets slowed down by an interrupt sufficiently to miss that moment, leading to another hour of sleep.

So code like this might break:
```
  initialize_interrups(); // interrupts will do all the heavy lifting
  loop {
    delay_us(u32::MAX); // sleep ~1hour
    do_some_hourly_maintenance();
  }
```

Reply:
I implemented a version that shouldn't have this issue in https://github.com/rp-rs/rp-hal/pull/614/commits/6aef40d957c6925e8039a97c50c7d0e0868297c1

While it looks a little bit ugly, it get optimized very well. For the common case of a constant delay, it gets compiled like this:
```
fn do_delay(delay: &mut dyn DelayUs<u32>) {
    delay.delay_us(500000);
}
```
⇓
```
rp2040_project_template::do_delay:
        ldr r0, .LCPI1_0 # address of the timelr register
        ldr r2, [r0]     # start = timelr
        ldr r1, .LCPI1_1 # delay value (500000)
.LBB1_1:
        ldr r3, [r0]     # now = timelr
        subs r2, r3, r2  # waited = now - start
        subs r1, r1, r2  # delay = delay - waited
        mov r2, r3       # started = now
        bhi .LBB1_1      # loop if subs resulted in a positive value
        bx lr
```